### ISSUE TITLE
Subscription field must return Async Iterable. Received: undefined

### ISSUE THREAD

**Issue by 4F2E4A2E** [apollographql/graphql-subscriptions#169]
We are trying to get the resolver for a subscription working in Apollo Server. The subscription works when it is a top-level field (i.e. directly under Subscription at the schema root).

However, if the subscription is contained within another type, I always get the error Subscription field must return Async Iterable. Received: undefined on the client's websocket connection -- the server's resolver is never executed.

i.e. this schema works:

```
type Subscription {
  postAdded: Post
}
```

**but this one does _not_:**

```
type Subscription {
  post: PostSubscription
}

type PostSubscription {
  postAdded: Post
}
```

My resolver for the second case looks like this, but I've tried a bunch of different variations with no success:

```
Subscription: {
  post: () => ({
    PostSubscription: {}
  })
},
PostSubscription: {
  postAdded: {
    subscribe: () => pubSub.asyncIterator(['postAdded'])
  }
}
```

Maybe relevant package versions:
```
"apollo-server-express": "1.3.6",
"apollo-server-module-graphiql": "1.3.4",
"graphql-subscriptions": "0.5.8",
"graphql": "0.13.2",
```

P.s.: Maybe unrelated, but just in case, this reminds me of https://github.com/yarax/swagger-to-graphql/commit/fa1000accd046b5db9f5d973e2e120c3f658f3b6#diff-6d186b954a58d5bb740f73d84fe39073L54

> Clone of: https://stackoverflow.com/questions/51862571/apollo-server-subscription-nested-within-a-type

**Comment by evseevnntmp** [comment]
Same trouble

**Comment by grantwwu** [comment]
Can you get a minimal reproducing example?

**Comment by GlennMatthys** [comment]
Minimal reproducing example: https://github.com/ECraneWorldwide/graphql-nested-subscriptions

**Comment by grantwwu** [comment]
I am having difficulty reproducing this.  I'm getting something like this instead:

```
{
  "data": {
    "foo": {
      "nested": null
    }
  },
  "errors": [
    {
      "message": "String cannot represent value: { fooObj: true }",
      "locations": [
        {
          "line": 3,
          "column": 5
        }
      ],
      "path": [
        "foo",
        "nested"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "stacktrace": [
            "TypeError: String cannot represent value: { fooObj: true }",
            "    at GraphQLScalarType.serializeString [as serialize] (/Users/grant.wu/foss/graphql-nested-subscriptions/node_modules/graphql/type/scalars.js:146:9)",
            "    at completeLeafValue (/Users/grant.wu/foss/graphql-nested-subscriptions/node_modules/graphql/execution/execute.js:641:37)",
            "    at completeValue (/Users/grant.wu/foss/graphql-nested-subscriptions/node_modules/graphql/execution/execute.js:587:12)",
            "    at completeValueCatchingError (/Users/grant.wu/foss/graphql-nested-subscriptions/node_modules/graphql/execution/execute.js:503:19)",
            "    at resolveField (/Users/grant.wu/foss/graphql-nested-subscriptions/node_modules/graphql/execution/execute.js:447:10)",
            "    at executeFields (/Users/grant.wu/foss/graphql-nested-subscriptions/node_modules/graphql/execution/execute.js:293:18)",
            "    at collectAndExecuteSubfields (/Users/grant.wu/foss/graphql-nested-subscriptions/node_modules/graphql/execution/execute.js:717:10)",
            "    at completeObjectValue (/Users/grant.wu/foss/graphql-nested-subscriptions/node_modules/graphql/execution/execute.js:707:10)",
            "    at completeValue (/Users/grant.wu/foss/graphql-nested-subscriptions/node_modules/graphql/execution/execute.js:598:12)",
            "    at completeValueCatchingError (/Users/grant.wu/foss/graphql-nested-subscriptions/node_modules/graphql/execution/execute.js:503:19)"
          ]
        }
      }
    }
  ]
}
```

I don't quite understand what the `foo` literal is supposed to be.

**Comment by RedShift1** [comment]
Which query are you trying to execute?

**Comment by grantwwu** [comment]
Are you asking me or OP?  I was trying
```
subscription{ foo { nested } }
```

**Comment by RedShift1** [comment]
I was asking you. Curious you are getting a different result, I'll try it myself again.

**Comment by RedShift1** [comment]
Nope, getting the exact same result:
```
subscription{foo{nested}}
```
```
{
  "error": {
    "message": "Subscription field must return Async Iterable. Received: undefined"
  }
}
```

![image](https://user-images.githubusercontent.com/3098010/46323693-ca7f5c00-c5ef-11e8-869a-c14b9756058c.png)

**Comment by grantwwu** [comment]
Let me try again tomorrow using `npm ci` instead of `npm install`

**Comment by RedShift1** [comment]
Were you able to reproduce it?

**Comment by grantwwu** [comment]
Sorry, more fires started at work.  I'll let you know when I get back to this...

**Comment by builtbyproxy** [comment]
Has there been any update on this?

**Comment by abelosorio** [comment]
I've been following the code and it seems that the resolver function is always `undefined` for some reason (that I can't understand).

I've read a lot of issues and posts related to this issue and I think the problem comes from GraphQL giving the schema to the transport.

ref https://github.com/apollographql/subscriptions-transport-ws/issues/480

Sorry if my comment makes no sense, just trying to help to find the solution. I really need to implement this feature using apollographql 2.

**Comment by yang4515** [comment]
Same trouble

**Comment by abelosorio** [comment]
**UPDATE: Sorry, my bad!!!**

I forgot the final export for my resolvers module...

Everything working perfectly now!

**Comment by builtbyproxy** [comment]
@abelosorio 	by final export are you talking about: 
```
schema {
	query: Query
	mutation: Mutation
	subscription: Subscription
}
```

or something else?

**Comment by builtbyproxy** [comment]
Ah! Fantastic thankyou very much for getting back to me @abelosorio

**Comment by grantwwu** [comment]
I have some time now and I'm a bit confused - @builtbyproxy @abelosorio were you working on the same issue as OP?  @RedShift1 are you still having issues?

**Comment by grantwwu** [comment]
Okay, I have managed to reproduce the issue locally.

**Comment by grantwwu** [comment]
Upon staring at these resolvers - it's not clear to me why this is actually supposed to work.  `foo` needs to be an object with a `resolve` and a `subscribe` field - instead right now it's returning a `{ fooObj: true }`.

I tried to fix up your resolvers into something that I think approximates what you want.  Here it is: https://github.com/grantwwu/graphql-nested-subscriptions/compare/grantwu_fixed~...grantwwu:grantwu_fixed

(Note that I also made some formatting changes, as well as updating your dependencies when I was ruling that out as a bug).

**Comment by SDLWebDev** [comment]
I somehow miss the answer of 4F2E4A2E's initial question.  Is there a support of subscriptions in subfields (non root subscriptions)? Do I oversee anything?

**Comment by grantwwu** [comment]
@4F2E4A2E never responded, so I'm not sure if their issue was resolved.

What do you mean by "subscriptions in subfields"?  A subscriptions are separate from queries and mutations, you can't mix them.

**Comment by SDLWebDev** [comment]
> The subscription works when it is a top-level field

If I understood @4F2E4A2E correctly, all non root / top level subscriptions are currently not working. He gave an exact and precise schema example. Is the described behaviour expected?

**Comment by grantwwu** [comment]
I don't think the given schema is supposed to work.  Subscriptions are a kind of operation, it doesn't make sense to put them as a type.

**Comment by hansololai** [comment]
I got here from a [stack overflow question](https://stackoverflow.com/questions/51862571/apollo-server-subscription-nested-within-a-type) and which I just answered. I want to just point out it is probably not a good design to try to nest the subscription fields under some object, (aka, not make these subscriptions just under Root subscription. Actually I would agree with @grantwwu that the schema should not work. Of course you could try to make it work, but there are issues with it. You can see more in my [answer with the quest](https://stackoverflow.com/questions/49803557/apollo-graphql-subscription-response-doesnt-handle-nested-queries/52683069#52683069)

**Comment by dunatron** [comment]
```
const { GraphQLServer } = require("graphql-yoga")
const Mutation = require("./resolvers/Mutation")
const Query = require("./resolvers/Query")
const Subscription = require("./resolvers/Subscription")
const db = require("./db")

// collect our graphql resolvers
const resolvers = {
  Query,
  Mutation,
  Subscription,
}

// create the graphql yoga server
function createServer() {
  return new GraphQLServer({
    typeDefs: "src/schema.graphql",
    resolvers: resolvers,
    resolverValidationOptions: {
      requireResolversForResolveType: false,
    },
    context: req => ({ ...req, db }),
  })
}

module.exports = createServer
```

**Comment by builtbyproxy** [comment]
@dunatron Could you please edit the above comment with a bit of context and such for anyone approaching it. For exmaple, all of a sudden we need to add [graphql-yoga][https://github.com/prisma/graphql-yoga] to our project... ?

**Comment by webgururobin** [comment]
Had the same problem, the subscribe function needs to return a value.

**Comment by iamcristos** [comment]
hello has this issue been fixed? currently having the same problem

**Comment by agborkowski** [comment]
@iamcristos more less check have you export subscription and you don't use nested types for subscription :) read that thread carefully ;)

**Comment by visu11290** [comment]
**Query :** 
Subscription field must return Async Iterable.

**Solution :**
When subscription not export in resolver module it will get above type error.

export default { Mutation, Query, Subscription };

**Comment by Adept-Rohan** [comment]
same trouble

**Comment by 4F2E4A2E** [comment]
@Adept-Rohan and any else, you're maybe better off, each of you creating an issue for this and referencing to this one hence this one cant be reopened.

**Comment by njnijesh** [comment]
If someone is facing this issue in production using typescript and not in development then it might because there is something wrong with exporting your resolvers.
```
import path from "path";
import { mergeResolvers } from "@graphql-tools/merge";
import { loadFilesSync } from "@graphql-tools/load-files";

const resolvers = loadFilesSync(path.join(__dirname, "/**/*.resolver.{ts}"));

export default mergeResolvers(resolvers);

``` 
In prooduction the files will be .js, so loadFilesSync will not be able to find the correct resolvers.
